[ESI] Software API creation infrastructure

teqdruid opened the pull request:
Build a model of the ESI system described in a `services.json` file. No 'backend' connectivity providers in this cut. Only integer types are supported for now.

teqdruid commented:
Since this is not strictly a "binding", I'm not sure if this is the right place for this sort of functionality. If not, I'm not sure where is.

mortbopet commented on frontends/PyCDE/src/pycde/system.py:
```diff
@@ -241,6 +242,15 @@ def createdb(self, primdb: PrimitiveDB = None):
+    """Build an ESI runtime API."""
+    if lang != "python":
+      raise ValueError(f"Language '{lang}' not supported")
+    b = PythonApiBuilder(
+        (self._output_directory / "services.json").open().read(),
+        (self._output_directory / "schema.capnp").open().read())
+    b.build(self.name, self._output_directory)
```
Nit: sanity check to ensure that these files exist.

mortbopet commented on lib/Bindings/Python/circt/dialects/esi.py:
```diff
@@ -4,3 +4,318 @@
+    # Create __init__.py and copy in the standard files.
+    init_file = libdir / "__init__.py"
+    init_file.touch()
+    common_file = libdir / "common.py"
+    common_file.write_text((__dir__ / "esi_runtime_common.py").read_text())
```
What is the reasoning behind copying a common file instead of including libraries distributed within the PyCDE package, seeing as that common file isn't parameterized? That would seem more natural and maintainable to me.

e.g.
- for python "pycde/esi/api/py/common.py"
- for c++ "pycde/esi/api/cpp/common.h, (cmake includes, ...), ..."

teqdruid replied:
These are runtimes which are meant to be distributed with the compiled image, not just used with the compiler.

mortbopet reviewed (commented):
Some comments.
Would also be good with a test to `FileCheck` the generated Python module.

mikeurbach reviewed (commented):
I haven't looked at this too deeply, but I had some thoughts on the implementation.

Rather than printf-ing Python from Python, would it be possible to write generic guts that use the service schema? I think it would be possible to write actual Python code that uses some metaprogramming to reflect out an API based on the json. Then we could ship this as a generic library instead of generating Python code.

Barring the above, if we are just going to printf Python from Python, maybe the ast module could help. It might be possible to write snippets of actual Python that are dumped with ast.unparse instead of snippets of Python in Python strings.

Anyway, just some thoughts. If you don't want to go this direction then this seems about right for generating the Python code.

teqdruid commented:
> Rather than printf-ing Python from Python, would it be possible to write generic guts that use the service schema? I think it would be possible to write actual Python code that uses some metaprogramming to reflect out an API based on the json. Then we could ship this as a generic library instead of generating Python code

So this is a design point I considered. It's what pycapnp does, and I don't like it. Mostly since it precludes any code inspection and type hinting. So it's impossible for an editor (or mypy) to provide autocompletion or automated checking.



> Barring the above, if we are just going to printf Python from Python, maybe the ast module could help. It might be possible to write snippets of actual Python that are dumped with ast.unparse instead of snippets of Python in Python strings

That's an interesting idea. I'll look into it, but my gut tells me that it'd be more complex than it's worth. Sometimes doing the dumb-shit thing is better then doing the fancy thing ;) .



> Would also be good with a test to `FileCheck` the generated Python module.

This would be a fragile test. Since testing this can only be done in the integration tests, I think we're better exercising the hell out of the functionality via cosim. So more integration tests will come when I add cosim support.

mikeurbach commented:
The argument for type completion and type checking makes sense to me. If the plan is to generate strings, maybe it would be a little more ergonomic with jinja2 or some form of templates instead of just a bunch of `write`s with Python string interpolation. It'd be nice if there were just templates for each API that get hydrated, instead of a bunch of Python and templates interspersed. But I haven't thought too much about the problem, and if a template engine can handle the kind of situations we have here.

teqdruid commented:
> > Barring the above, if we are just going to printf Python from Python, maybe the ast module could help. It might be possible to write snippets of actual Python that are dumped with ast.unparse instead of snippets of Python in Python strings
> 
> That's an interesting idea. I'll look into it, but my gut tells me that it'd be more complex than it's worth. Sometimes doing the dumb-shit thing is better then doing the fancy thing ;) .

Turns out `unparse` was only introduced in python 3.9... I couldn't find a well-documented python library with similar functionality. This also looks too complicated to be worth it, but I may be wrong long term.

teqdruid commented:
> It'd be nice if there were just templates for each API that get hydrated, instead of a bunch of Python and templates interspersed. But I haven't thought too much about the problem, and if a template engine can handle the kind of situations we have here

This is another design point I considered. It's got two problems as I see it: 1) (minor) it's an extra dependency and may pull in many more dependencies. 2) The truly ugly parts have logic sufficiently complex that I don't think any template engine could handle it, so it'd be used as a replacement engine only. And with format-strings, python isn't terrible. At least not terrible enough to justify the extra hop into a template language. I may have gone a bit far with the inline string computation, which obscures the seeing the overall structure.

For instance, this code isn't _too_ bad:
```py
    for port in decl["ports"]:
      pn = port['name']
      if "to-server-type" in port:
        write(f"""
  def {pn}_read_any(self):
    for p in self.{pn}:
      rc = p.read(block=False)
      if rc is not None:
        return rc
    return None""")
```

mikeurbach commented:
> For instance, this code isn't too bad

I don't think that example is much different than a jinja template on its own, but my gripe looking at this PR is you can't just look at all of the code to be generated top-to-bottom. There's lots of little snippets like that with intervening Python, which makes it hard to see which class your part of, and things like that. If it was a template, it would be a big block of template with code interspersed, rather than a big block of code with templates interspersed.

> The truly ugly parts have logic sufficiently complex that I don't think any template engine could handle it

I'm not sure about that, jinja2 has a pretty extensive language that supports a lot of the same constructs you'd have in Python. For example: https://jinja.palletsprojects.com/en/3.0.x/templates/#list-of-control-structures

I don't feel too strongly about this, but I wanted to try and make the case for a template engine since they're built for exactly this use case.

teqdruid commented:
@mikeurbach I moved over to Jinja2. It's pretty ugly IMO, but you're right -- it's more legible than the Python version. It's certainly more compact. Moved in into PyCDE since I couldn't justify adding the Jinja2 dependency to the circt python bindings.

mikeurbach reviewed (commented):
Nice, thanks for playing with jinja2. I personally think this is an improvement, and this looks like a cleaner separation of data gathering versus templating to me. I can give a deeper review in a bit.

teqdruid commented:
@mikeurbach Yeah, after sleeping it does look nicer. I'm also planning on using the same format for C++ API when I do that. Is there some way to re-use the jinja2 logic for both? I'm pretty sure it'll be identical but the python code will just be c++. I would imagine not, so I'll just have to copy it.

mikeurbach commented:
> Is there some way to re-use the jinja2 logic for both?

That would be great... though I don't know offhand. Maybe there can be some kind of indirection with macros? In the template you call a macro like `call render` but you use something to selectively import either the `render` for python or the `render` for c++.
